### PR TITLE
Add preprocess & capability for custom dataset for mt

### PR DESCRIPTION
**Pull request by FrostML**
### PR types

New features

### PR changes

Others

### Description

Add preprocess & capability for custom dataset for mt.

### CODE REVIEW

File: examples/machine_translation/preprocessor/preprocessor.py [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+    parser.add_argument("--eos_token",
+                        default="</s>",
+                        type=str,
+                        help="eos_token. ")
+    parser.add_argument(
+        "--pad_token",
+        default=None,
+        type=str,
+        help=
+        "The token used for padding. If it's None, the bos_token will be used. Defaults to None. "
```
Comment (gpengzhi) [inline examples/machine_translation/preprocessor/preprocessor.py]: Why is `None` used as default setting? Maybe we shoulde use `<pad>` as default instead.

File: examples/machine_translation/transformer/README.md [hunk 2]
```diff
@@ -85,11 +73,74 @@ python -m paddle.distributed.launch --gpus "0,1,2,3,4,5,6,7" train.py --config .
...
+python train.py \
+    --config configs/transformer.base.yaml \
+    --train_file ${DATA_DEST_DIR}/train.de-en.de ${DATA_DEST_DIR}/train.de-en.en \
+    --dev_file ${DATA_DEST_DIR}/dev.de-en.de ${DATA_DEST_DIR}/dev.de-en.en \
+    --src_vocab ${DATA_DEST_DIR}/dict.en.txt \
+    --trg_vocab ${DATA_DEST_DIR}/dict.de.txt \
+    --bos_token "<s>" \
+    --eos_token "</s>" \
+    --unk_token "<unk>"
+```
```
Comment (gpengzhi) [inline examples/machine_translation/transformer/README.md:99]: `pad_token` is not supported in the training step as well as the following steps?

File: examples/machine_translation/transformer/README.md [hunk 3]
```diff
@@ -115,6 +200,29 @@ python export_model.py --config ./configs/transformer.base.yaml
...
+同样，因为模型导出会用到模型的词表等信息，所以如果是**自定义数据集**，仍需要传入所使用的词表。
+
+``` bash
+DATA_DEST_DIR=${PATH_TO_PADDLENLP}/PaddleNLP/examples/machine_translation/data/iwslt14.tokenized.de-en/
+
+python export_model.py \
+    --config ./configs/transformer.base.yaml \
+    --src_vocab ${DATA_DEST_DIR}/dict.en.txt \
+    --trg_vocab ${DATA_DEST_DIR}/dict.de.txt \
+    --bos_token "<s>" \
```
Comment (gpengzhi) [inline examples/machine_translation/transformer/README.md:215]: The special tokens are also required when converting to static model?

File: examples/machine_translation/transformer/README.md [hunk 4]
```diff
@@ -85,11 +73,74 @@ python -m paddle.distributed.launch --gpus "0,1,2,3,4,5,6,7" train.py --config .
...
+    --eos_token "</s>" \
+    --unk_token "<unk>"
+```
+
+`train.py` 脚本中，各个参数的含义如下：
+
+* `--config`: 指明所使用的 Transformer 的 config 文件，包括模型超参、训练超参等，默认是 `transformer.big.yaml`。即，默认训练 Transformer Big 模型。
+* `--data_dir`: 指明训练需要的数据集的路径。无需提供不同的 train、dev 和 test 文件具体的文件名，会自动根据 `--src_lang` 和 `--trg_lang` 指定的语言进行构造。train、dev 和 test 默认的文件名分别为 [train|dev|test].{src_lang}-{trg_lang}.[{src_lang}|{trg_lang}]。且 `--data_dir` 设置的优先级会高于后面提到的 `--train_file`，`--dev_file` 和 `--test_file` 的优先级。
+  * `--src_lang`(`-s`): 指代翻译模型的源语言。比如 `de` 表示德语，`en` 表示英语，`fr` 表示法语等等。和数据集本身相关。
```
Comment (gpengzhi) [inline examples/machine_translation/transformer/README.md:105]: Is the indentation here as expected?

File: examples/machine_translation/transformer/README.md [hunk 5]
```diff
@@ -162,9 +272,88 @@ python -m paddle.distributed.launch --gpus="0,1,2,3,4,5,6,7" train.py --config .
...
+* `--distributed`:（**多卡训练需要**）指明是否是使用 fleet 来启动多卡。若设置，则使用 fleet 启动多卡。具体使用方式如下。
+
+##### PE 的方式启动单机多卡：
+``` shell
+cd static/
+export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
+python train.py \
+    --config ../configs/transformer.base.yaml \
+    --train_file ${DATA_DEST_DIR}/train.de-en.de ${DATA_DEST_DIR}/train.de-en.en \
+    --src_vocab ${DATA_DEST_DIR}/dict.en.txt \
```
Comment (gpengzhi) [inline examples/machine_translation/transformer/README.md:333]: `dev_file` is not supported in the static mode?

File: examples/machine_translation/transformer/deploy/python/inference.py [hunk 6]
```diff
@@ -52,13 +66,20 @@ def parse_args():
...
+        "The dir of train, dev and test datasets. If data_dir is given, train_file and dev_file and test_file will be replaced by data_dir/[train|dev|test].\{src_lang\}-\{trg_lang\}.[\{src_lang\}|\{trg_lang\}]. "
+    )
     parser.add_argument(
         "--test_file",
         nargs='+',
         default=None,
         type=str,
         help=
-        "The file for testing. Normally, it shouldn't be set and in this case, the default WMT14 dataset will be used to process testing."
+        "The files for test, including [source language file, target language file]. If it's None, the default WMT14 en-de dataset will be used. "
```
Comment (gpengzhi) [inline examples/machine_translation/transformer/deploy/python/inference.py]: Why do we need target language file during testing?

File: examples/machine_translation/transformer/README.md [hunk 7]
```diff
@@ -99,9 +150,43 @@ python predict.py --config ./configs/transformer.base.yaml
...
+    --trg_vocab ${DATA_DEST_DIR}/dev.de-en.en \
+    --bos_token "<s>" \
+    --eos_token "</s>" \
+    --unk_token "<unk>"
+```
+
+以下是各个参数的含义：
+
+* `--config`: 指明所使用的 Transformer 的 config 文件，包括模型超参、训练超参等，默认是 `transformer.big.yaml`。即，默认训练 Transformer Big 模型。
+* `--data_dir`: 指明训练需要的数据集的路径。无需提供不同的 train、dev 和 test 文件具体的文件名，会自动根据 `--src_lang` 和 `--trg_lang` 指定的语言进行构造。train、dev 和 test 默认的文件名分别为 [train|dev|test].{src_lang}-{trg_lang}.[{src_lang}|{trg_lang}]。且 `--data_dir` 设置的优先级会高于后面提到的 `--train_file`，`--dev_file` 和 `--test_file` 的优先级。
```
Comment (gpengzhi) [inline examples/machine_translation/transformer/README.md:179]: Do we need this feature during testing?

File: examples/machine_translation/transformer/README.md [hunk 8]
```diff
@@ -99,9 +150,43 @@ python predict.py --config ./configs/transformer.base.yaml
...
+  * `--src_lang`(`-s`): 指代翻译模型的源语言。比如 `de` 表示德语，`en` 表示英语，`fr` 表示法语等等。和数据集本身相关。
+  * `--trg_lang`(`-t`): 指代翻译模型的目标语言。比如 `de` 表示德语，`en` 表示英语，`fr` 表示法语等等。和数据集本身相关。
+* `--test_file`: 指明训练所需要的 `test` 验证集的数据集的路径。若没有提供 `--data_dir` 或是需要特别指明训练数据的名称的时候指定。指定的方式为，可以是一组平行语料的源语言和目标语言，依次两个文件的路径和名称，`--test_file ${SOURCE_LANG_FILE} ${TARGET_LANG_FILE}`，也可以只转入源语言的文件。比如，`--test_file ${DATA_DEST_DIR}/test.de-en.de ${DATA_DEST_DIR}/test.de-en.en` 或者 `--test_file ${DATA_DEST_DIR}/test.de-en.de`。
```
Comment (gpengzhi) [inline examples/machine_translation/transformer/README.md]: Why do we need to provide parallel files during testing? Is there any specific scenraio requires such feature?

**Comment by gpengzhi** [comment]
To verify the correctness of our changes, could you reproduce the experimental results on the IWSLT14 de->en dataset by following the [tutorial](https://github.com/facebookresearch/fairseq/tree/main/examples/translation#iwslt14-german-to-english-transformer)?

We could add a new config file (`transformer.iwslt14.yaml`) to exactly match the model (`trainsformer_iwslt_de_en`) and training configurations used in the `fairseq` example. Since we (`fairseq` and `paddlenlp`) share the same preprocessing script (`prepare-iwslt14.sh`), we should achieve similiar results. 

We could run this experiment on a single GPU, and it won't take too much time to converge. The final BLEU score should be around 34.85.

File: examples/machine_translation/preprocessor/preprocessor.py (hunk 1 above)
Comment (FrostML) [reply examples/machine_translation/preprocessor/preprocessor.py]: This is for compatibility concern. 
In the former example, Transformer used `bos_token` as `pad_token`. Hence, during training for now, if `pad_token` is `None`, the `bos_token` will be used for padding. Hence, in preprocessing, we keep this behavior by setting default `pad_token` is `None`.

File: examples/machine_translation/transformer/README.md (hunk 4 above)
Comment (FrostML) [reply examples/machine_translation/transformer/README.md:105]: Yes. `--src_lang` and `--trg_lang` are only used with `--data_dir` to specify the train, dev and test data file.

File: examples/machine_translation/transformer/README.md (hunk 8 above)
Comment (FrostML) [reply examples/machine_translation/transformer/README.md]: It's not necessary. This is for compatibility concern. I will delete this description in doc. Thanks for pointing out.

File: examples/machine_translation/transformer/README.md (hunk 3 above)
Comment (FrostML) [reply examples/machine_translation/transformer/README.md:215]: This is for adapt the vocabulary size and get special token id.  
`<unk>` is not required. I'll delete this.

File: examples/machine_translation/transformer/README.md (hunk 5 above)
Comment (FrostML) [reply examples/machine_translation/transformer/README.md:333]: True. There is still no validation step in static mode train.

File: examples/machine_translation/transformer/deploy/python/inference.py (hunk 6 above)
Comment (FrostML) [reply examples/machine_translation/transformer/deploy/python/inference.py]: I'll fix the `help`. It not necessary to provide target language file in testing.

File: examples/machine_translation/transformer/README.md (hunk 2 above)
Comment (FrostML) [reply examples/machine_translation/transformer/README.md:99]: It should support `pad_token`. I'll fix the doc and train/predict. Thanks.

File: examples/machine_translation/transformer/README.md (hunk 7 above)
Comment (FrostML) [reply examples/machine_translation/transformer/README.md:179]: Not necessary, just being the same as train scripts.

File: examples/machine_translation/README.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+除了 iwslt14 德英翻译数据集外，我们也提供了其他的 shell 脚本完成数据下载处理，其中包括了 WMT14 英德翻译数据和 WMT14 英法翻译数据。
+
+``` bash
+# WMT14 英德翻译的数据下载、处理
+bash prepare-wmt14en2de.sh
+# WMT14 英法翻译的数据下载、处理
+bash prepare-wmt14en2fr.sh
+```
+
+完成数据处理之后，同样也可以采用上文提到的预处理方式获取词表，完成预处理。以下再以 WMT14 EN-DE 翻译数据集预处理为例：
```
Comment (gpengzhi) [inline examples/machine_translation/README.md]: If we decide to use WMT14 en-de as an example, why do we need to mention IWSLT14 de-en above?

File: examples/machine_translation/transformer/README.md [hunk 10]
```diff
@@ -85,11 +73,76 @@ python -m paddle.distributed.launch --gpus "0,1,2,3,4,5,6,7" train.py --config .
...
 
+### 使用自定义数据集进行训练
+
+自定义数据集与内置数据集训练的方式基本上是一致的，不过需要额外提供数据文件的路径。可以参照以下文档。
+
+#### 单机单卡
+
+本示例这里略去自定义数据下载、处理的步骤，如果需要，可以参考前页文档 [使用自定义翻译数据集](../README.md)。
+
+本示例以处理好的 iwslt14 数据为例。
```
Comment (gpengzhi) [inline examples/machine_translation/transformer/README.md]: If we want to use IWSLT14 as an example, transformer-base maybe too large for this dataset. To keep the instruction consistent, we could simply use WMT14 en-de as the example.

File: examples/machine_translation/README.md (hunk 9 above)
Comment (FrostML) [reply examples/machine_translation/README.md]: Thanks. This should be fixed.

File: examples/machine_translation/transformer/README.md (hunk 10 above)
Comment (FrostML) [reply examples/machine_translation/transformer/README.md]: Fixed. Thanks.

**Comment by gpengzhi** [comment]
Do we plan to merge this into `master` branch?

File: examples/machine_translation/README.md [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+#### 使用内置已经处理完成数据集
+
+内置的处理好的数据集是基于公开的数据集：WMT 数据集。
+
+WMT 翻译大赛是机器翻译领域最具权威的国际评测大赛，其中英德翻译任务提供了一个中等规模的数据集，这个数据集是较多论文中使用的数据集，也是 Transformer 论文中用到的一个数据集。我们也将 [WMT'14 EN-DE 数据集](http://www.statmt.org/wmt14/translation-task.html) 作为示例提供。
+
+可以编写如下代码，即可自动载入处理好的上述的数据，对应的 WMT14 EN-DE 的数据集将会自动下载并且解压到 `~/.paddlenlp/datasets/WMT14ende/`。
+
+``` python
+datasets = load_dataset('wmt14ende', splits=('train', 'dev'))
```
Comment (guoshengCS) [inline examples/machine_translation/README.md:29]: 这里还是之前paddlenlp中的`load_dataset`和Dataset吗，统一更新使用HF的`load_dataset`和Dataset吧。下面新增的那个`LanguagePairDataset`也是能支持这个吧，但是好像没看到`LanguagePairDataset`在这个PR里的使用。

另外针对`LanguagePairDataset`，同 @sijunhe 的comment https://github.com/PaddlePaddle/PaddleNLP/pull/3698#discussion_r1022249549 ，也如之前沟通如果放在package里作为通用API的话不建议load language_pair直接返回一个wmt En-De BPE的数据集出来，或者是否可以通过类似glue的方式以name指定来实现返回特定的parallel corpora，这样算是对HF那些具体翻译数据集的更高一层接口，请 @sijunhe 再看看这个PR呢

**Comment by FrostML** [comment]
@guoshengCS

File: examples/machine_translation/README.md [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+* 基于 `mosesdecoder/scripts/training/clean-corpus-n.perl` 完成数据的清洗；
+* 使用 `subword-nmt/subword_nmt/learn_bpe.py` 完成 bpe 的学习；
+
+此外，基于学到的 bpe code 进行分词的操作目前提供了两种选项，其一是，可以在以上的 shell 脚本中处理完成，使用以下的工具：
+
+* 使用 `subword-nmt/subword_nmt/apply_bpe.py` 完成分词工作。
+
+其二，也可以直接在后面的 `preprocessor/preprocessor.py` 脚本中，指明 `--apply_bpe` 完成分词操作。
+
+而在预处理 `preprocessor/preprocessor.py` 脚本中，则提供词表构建，数据集文件整理，甚至于 bpe 分词的功能（bpe 分词过程可选）。最后获取的处理完成的 train，dev，test 数据可以直接用于后面 Transformer 模型的训练、评估和推理中。
```
Comment (guoshengCS) [inline examples/machine_translation/README.md]: > `preprocessor/preprocessor.py` 支持了在机器翻译中常见的数据预处理方式

建议把preprocessor.py的功能放在上面对其介绍的地方，避免看了一堆到最后了才知道preprocessor.py能干啥

File: examples/machine_translation/README.md (hunk 11 above)
Comment (FrostML) [reply examples/machine_translation/README.md:29]: 当前 PR 已全部替换为 `LanguagePairDataset`，这个可以在 reader 看到。

这个和 glue 不一样，glue 指明了不同的 Dataset 能够返回出数据，这里指定任意的翻译方向还是得提供处理好的数据。返回 en-de 只是作为默认的返回结果，保证不会报错，可以置为空，抛出异常处理，只是这样会影响使用体验

替换内置数据集 Dataset 载入方式还是等 HF WMT Dataset 接入之后再替换文档吧，这里的 language pair 预期是用于自定义数据并且和前面的预处理连接，内置的 en-de 翻译 Dataset 还是用之前的。

File: examples/machine_translation/README.md (hunk 12 above)
Comment (FrostML) [reply examples/machine_translation/README.md]: Done.

File: examples/machine_translation/preprocessor/prepare-iwslt14.sh [hunk 13]
```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env bash
+# Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Comment (sijunhe) [inline examples/machine_translation/preprocessor/prepare-iwslt14.sh]: 这里所有的license都应该是2022

File: examples/machine_translation/README.md [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+
+### 环境依赖
+
+使用当前机器翻译示例，需要额外安装配置以下环境：
+
+* attrdict
+* pyyaml
+* subword_nmt
+* sacreBLEU
+* fastBPE (可选，若不使用 preprocessor.py 的 bpe 分词功能可以不需要)
```
Comment (sijunhe) [inline examples/machine_translation/README.md]: 这里是不是通过requirements.txt来指定更合理？

Comment (FrostML) [reply examples/machine_translation/README.md]: 不合适，作为可选依赖，不强制安装

Comment (FrostML) [reply examples/machine_translation/README.md]: 已新增 requirements.txt 将非可选依赖加入。THX

Comment (sijunhe) [reply examples/machine_translation/README.md]: 我的意思并不是加入主干的requirements.txt, 而是`examples/machine_translation/requirements.txt`, 这样子用户可以直接`pip install -r requirements.txt`

File: examples/machine_translation/preprocessor/prepare-iwslt14.sh (hunk 13 above)
Comment (FrostML) [reply examples/machine_translation/preprocessor/prepare-iwslt14.sh]: DONE. THX.

File: examples/machine_translation/README.md (hunk 14 above)
Comment (FrostML) [reply examples/machine_translation/README.md]: 是的，目前这样加上了

**Comment by FrostML** [comment]
@guoshengCS

File: examples/machine_translation/README.md (hunk 11 above)
Comment (guoshengCS) [reply examples/machine_translation/README.md:29]: 这个`LanguagePairDataset ` API加上 `logger.warning` 说明 subject to change 吧

**Review by guoshengCS — DISMISSED** [review]

**Review by guoshengCS — APPROVED** [review]

**Pull request merged**